### ISSUE TITLE
Windows 1903 / WSL 1 "Permission denied" from UPX check

### ISSUE THREAD

**Issue by asantoni** [pyinstaller/pyinstaller#4352]
I have a weird problem that is probably not a PyInstaller bug, but other people are likely to run into this problem and I have a patch to share that fixes it. I'm putting this here so people can find it from Google. (Feel free to close this if it's not considered a bug.)

On Windows 10 with the 1903 update, PyInstaller in WSL (version 1) fails to run, due to a "Permission denied" error when trying to detect `upx`. The check for the `upx` runs even if you pass `--noupx` and/or disable upx in your .spec file, so there's no way around this.

It's not clear to me why the "Permission denied" error happens. I've never had `upx` on this system, and all that I changed was upgrading from Windows 10 1809 to 1903, so I suspect it might be some kind of WSL 1 syscall bug that was introduced. I'm running Ubuntu 18.04 in WSL 1.

The exact error I get is:

```
$ pyinstaller --noupx  --noconfirm project.spec  --log-level=DEBUG
49 INFO: PyInstaller: 3.5
50 INFO: Python: 3.6.7
51 INFO: Platform: Linux-4.4.0-18362-Microsoft-x86_64-with-Ubuntu-18.04-bionic
51 DEBUG: Testing for UPX ...
----------------------------------------
Error running 'upx -V':
[Errno 13] Permission denied: 'upx'
----------------------------------------
Error: Executing command failed!
```

I have `upx` executable on my system, in any of my PATH directories.

My specfile is:

```
block_cipher = None

a = Analysis(['bin/project'],
             pathex=['/mnt/c/project'],
             binaries=[],
             datas=[('README.md', '.')],
             hiddenimports=[],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher,
             noarchive=False)
pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)
exe = EXE(pyz,
          a.scripts,
          a.binaries,
          a.zipfiles,
          a.datas,
          [],
          name='project',
          debug=False,
          bootloader_ignore_signals=False,
          strip=False,
          upx=False, # This flag doesn't actually prevent detection of the UPX executable
          upx_exclude=[],
          runtime_tmpdir=None,
          console=True )
```

A `strace` of the `pyinstaller` run shows that it checks each PATH directory, but then somehow ends up "Permission denied" error. I don't see the syscall that's returning that in the `strace` log, but I must be missing it somehow. It's gotta be in here somewhere:

```
[pid  1119] execve("/mnt/c/<redacted>/cmd/upx", ["upx", "-V"], [/* 20 vars */]) = -1 ENOENT (No such file or directory)
[pid  1119] execve("/mnt/c/<redacted>/bin/upx", ["upx", "-V"], [/* 20 vars */]) = -1 ENOENT (No such file or directory)
[pid  1119] execve("/mnt/c/<redacted>/upx", ["upx", "-V"], [/* 20 vars */]) = -1 ENOENT (No such file or directory)
[pid  1119] execve("/mnt/c/<redacted>/upx", ["upx", "-V"], [/* 20 vars */]) = -1 ENOENT (No such file or directory)
[pid  1119] execve("/mnt/c/Users/<redacted>/upx", ["upx", "-V"], [/* 20 vars */]) = -1 ENOENT (No such file or directory)
[pid  1119] execve("/mnt/c/<redacted>/bin/upx", ["upx", "-V"], [/* 20 vars */]) = -1 ENOENT (No such file or directory)
[pid  1119] execve("/mnt/c/<redacted>/upx", ["upx", "-V"], [/* 20 vars */]) = -1 ENOENT (No such file or directory)
[pid  1119] execve("/mnt/c/Users/<redacted>/AppData/<redacted>/upx", ["upx", "-V"], [/* 20 vars */]) = -1 ENOENT (No such file or directory)
[pid  1119] execve("/snap/bin/upx", ["upx", "-V"], [/* 20 vars */]) = -1 ENOENT (No such file or directory)
[pid  1119] write(6, "OSError:", 8 <unfinished ...>
[pid  1115] <... read resumed> "OSError:", 50000) = 8
[pid  1119] <... write resumed> )       = 8
[pid  1115] read(5,  <unfinished ...>
[pid  1119] write(6, "d", 1)            = 1
[pid  1115] <... read resumed> "d", 50000) = 1
[pid  1119] write(6, ":", 1 <unfinished ...>
[pid  1115] read(5,  <unfinished ...>
[pid  1119] <... write resumed> )       = 1
[pid  1115] <... read resumed> ":", 50000) = 1
[pid  1119] exit_group(255)             = ?
[pid  1115] read(5,  <unfinished ...>
[pid  1119] +++ exited with 255 +++
[pid  1115] <... read resumed> "", 50000) = 0
[pid  1115] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=1119, si_uid=1000, si_status=255, si_utime=0, si_stime=0} ---
[pid  1115] close(5)                    = 0
[pid  1115] wait4(1119, [{WIFEXITED(s) && WEXITSTATUS(s) == 255}], 0, NULL) = 1119
[pid  1115] close(3)                    = 0
[pid  1115] write(2, "----------------------------------------\n", 41----------------------------------------
) = 41
[pid  1115] write(2, "Error running 'upx -V':\n", 24Error running 'upx -V':
) = 24
[pid  1115] write(2, "[Errno 13] Permission denied: 'upx'\n", 36[Errno 13] Permission denied: 'upx'
) = 36
[pid  1115] write(2, "----------------------------------------\n", 41----------------------------------------
) = 41
[pid  1115] munmap(0x7f5648cb0000, 262144) = 0
[pid  1115] write(2, "Error: Executing command failed!\n", 33Error: Executing command failed!
) = 33
[pid  1115] rt_sigaction(SIGINT, {SIG_DFL, [], SA_RESTORER, 0x7f564c03ef20}, {0x62ff10, [], SA_RESTORER, 0x7f564c03ef20}, 8) = 0
[pid  1115] sigaltstack(NULL, {ss_sp=0x12f6fc0, ss_flags=0, ss_size=8192}) = 0
[pid  1115] sigaltstack({ss_sp=NULL, ss_flags=SS_DISABLE, ss_size=0}, NULL) = 0
[pid  1115] futex(0x7f564a3f682c, FUTEX_WAKE_PRIVATE, 2147483647) = 0
[pid  1115] exit_group(1)               = ?
[pid  1115] +++ exited with 1 +++
<... wait4 resumed> [{WIFEXITED(s) && WEXITSTATUS(s) == 1}], 0, NULL) = 1115
rt_sigaction(SIGINT, {SIG_DFL, [], SA_RESTORER, 0x7f1ffb1fef20}, {0x7f1ffbe54030, [], SA_RESTORER, 0x7f1ffb1fef20}, 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=1115, si_uid=1000, si_status=1, si_utime=0, si_stime=0} ---
wait4(-1, 0x7fffddd1e350, WNOHANG, NULL) = -1 ECHILD (No child processes)
rt_sigreturn({mask=[]})                 = 0
rt_sigprocmask(SIG_BLOCK, [CHLD], [], 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
exit_group(1)                           = ?
+++ exited with 1 +++
```

The solution / workaround I found is to hack `compat.py` in the PyInstaller package, adding the following lines after 414:

```
    if raise_ENOENT and e.errno == errno.EACCESS:
        raise
```

Basically, the search for `upx` uses exec_command from the `compat.py` file, which catches OSErrors, but caused the application to abort if it gets anything other than an `errno.ENOENT`. This is why the "Permission denied" error is fatal. My workaround just makes it raise the OSError exception instead of aborting, and the rest of the UPX detection logic copes with it and correctly knows that UPX isn't installed on my system. (This patch is a hack, and I don't recommend committing it. It's just here in case anyone needs to urgently fix their PyInstaller.)

**Comment by daniel-a-melo** [comment]
I can confirm that the same issue occurs with WSL2 on Windows 10 1909. 

As of PyInstaller 4.5.1 the fix suggested by @asantoni  should be applied to `compat.py` line 336

**Comment by ioef** [comment]
Hello.
Since I was facing the same problem (on a Linux Environment) I was also searching for a solution. Something that worked for me was to introduce the **upx-dir** argument in the command indicating a default path. Since I am not using upx, I presume any path will do:

$ pyinstaller --noupx **--upx-dir=/usr/local/share/** --noconfirm project.spec  --log-level=DEBUG

The above completely eliminated the error messages like 

"Error running 'upx -V':
[Errno 13] Permission denied: 'upx'"

which gave their place to the "INFO: UPX is not available".

Hope this helps someone in the future, to avoid blood, sweat and tears.

**Comment by rokm** [comment]
The UPX check has been reworked in #7707 (PyInstaller 5.13.0), so this should not be a problem on contemporary PyInstaller versions anymore.
